File: android/app/src/main/kotlin/games/yandex/wrap/network/GameDetailEndpointService.kt
package games.yandex.wrap.network

import games.yandex.wrap.network.dtos.GetGameResponseDto

class GameDetailEndpointService(
    private val networkService: NetworkService? = null,
) {
    suspend fun detail(appId: Long): GetGameResponseDto = requireNetworkService().execute(GameDetailRequest(appId))

    private fun requireNetworkService(): NetworkService =
        requireNotNull(networkService) { "NetworkService is required to execute game detail requests." }
}

data class GameDetailRequest(
    private val appId: Long,
) : Request<GetGameResponseDto> {
    override val method = HttpMethod.Post
    override val path = "/games/api/catalogue/v2/get_game"
    override val serializer = GetGameResponseDto.serializer()
    override val jsonBody = json("appID" to appId, "format" to "app")
}

File: ios/UGames/Network/GameDetailEndpointService.swift
import Foundation

struct GameDetailEndpointService {
    let networkService: NetworkService

    func detail(appId: Int64) async throws -> GetGameResponseDTO {
        try await networkService.execute(GameDetailRequest(appId: appId))
    }
}

struct GameDetailRequest: Request {
    typealias DTO = GetGameResponseDTO

    let appId: Int64

    var method: HTTPMethod { .post }
    var path: String { "/games/api/catalogue/v2/get_game" }
    var body: Data? {
        json(["appID": appId, "format": "app"])
    }
}
